sdk/rust/oak/src/lib.rs

```diff
@@ -462,3 +461,69 @@ pub fn run_event_loop<T: crate::io::Decodable, N: Node<T>>(mut node: N, in_handl
         }
     }
 }
+
+/// Register a new node entrypoint.
+///
+/// This registers the entrypoint name and the expression used to construct the
+/// node instance.
+///
+/// ```
+/// #[derive(Default)]
+/// struct DummyNode;
+///
+/// impl oak::grpc::OakNode for DummyNode {
+///     // ...
+///     # fn invoke(&mut self, _: &str, _: &[u8], _: oak::grpc::ChannelResponseWriter) {
+///     #     unimplemented!()
+///     # }
+/// }
+///
+/// oak::entrypoint!(dummy => DummyNode::default());
+///
+/// # fn main() {}
+/// ```
+///
+/// Entrypoints need to be declared in the global scope (as opposed to for example
+/// in a function body). Is recommended but not required to define all entrypoints of your
+/// module in `lib.rs`.
+///
+/// If instantiating your node requires some setup, it is possible to do that in the node
+/// expression too:
+/// ```
+/// # #[derive(Default)]
+/// # struct DummyNode;
+/// #
+/// # impl oak::grpc::OakNode for DummyNode {
+/// #     fn invoke(&mut self, _: &str, _: &[u8], _: oak::grpc::ChannelResponseWriter) {
+/// #         unimplemented!()
+/// #     }
+/// # }
+/// #
+/// oak::entrypoint!(its_complicated => {
+///     oak_log::init_default();
+///     DummyNode::default()
+/// });
+/// #
+/// # fn main() {}
+/// ```
+#[macro_export]
+macro_rules! entrypoint {
+    ($name:ident => $node:expr) => {
```

Comment:
I guess the fact that the name is an ident makes sense, since it's used to generate the exported fn, but somehow I would have expected the first arg to be a string.

@wildarch have you tried both versions, and did you prefer this one? Or was there a technical limitation preventing the string version from working?

Also curious to hear opinions from @project-oak/core (@daviddrysdale and @blaxill in particular)

Reply:
I haven't tried a version with a string parameter, but I would think macro hygiene prevents us from going that route. Generating fresh identifiers within a macro_rules! is impossible as far as I know, they have to be passed in as parameters.

Reply:
Good point; though I note that it's not necessarily the case that a Rust ident maps to a valid Wasm exported symbol name (for all I know there is no strict implication either way).

But anyways, given that Rust macros are indeed hygienic, perhaps this is the only way to bind an identifier.